File: GameOfPureStrategy_C/GOPSlocal/src/player.c
#include "player.h"
#include "card.h"
#include <stdio.h>

void initPlayer(Player *player) {
    player->handSize = 0;
    player->points = 0;
    for (int i = 1; i <= NUM_CARDS; i++) {
        Card card = createCard(i);
        addCard(player, card);
    }
}

int removeCard(Player *player, Card card) {
    int index = indexOf(player, card);
    if (index < 0 || index >= player->handSize) {
        return 0;
    }
    for (int i = index; i < player->handSize - 1; i++) {
        player->hand[i] = player->hand[i + 1];
    }
    player->handSize--;
    return 1;
}

void addCard(Player *player, Card card) {
    if (player->handSize < NUM_CARDS) {
        player->hand[player->handSize] = card;
        player->handSize++;
    }
}

int indexOf(Player *player, Card card) {
    for (int i = 0; i < player->handSize; i++) {
        if (card.rank == player->hand[i].rank) {
            return i;
        }
    }
    return -1;
}

void displayHand(Player *player) {
    for (int i = 0; i < player->handSize; i++) {
        printCard(&player->hand[i]);
        printf(" ");
    }
    printf("\n");
}

void addPoints(Player *player, int points) {
    player->points += points;
}

File: TheFoxInTheForest_Rust/common/src/player.rs
use std::io;
use std::io::Write;
use crate::card::*;

pub struct Player {
    hand: Vec<Card>,
    tricks: usize,
    score: usize
}

impl Player {
    pub fn new() -> Self {
        Player {
            hand: Vec::new(),
            tricks: 0,
            score: 0
        }
    }

    pub fn draw_card(&mut self, card: Card) {
        self.hand.push(card);
        self.sort_hand();
    }

    fn sort_hand(&mut self) {
        self.hand.sort();
    }

    pub fn display_hand(&self) {
        for card in &self.hand {
            print!("{} ", card.to_string());
        }
        io::stdout().flush().unwrap();
    }

    pub fn remove_card(&mut self, card: &Card) {
        if let Some(pos) = self.hand.iter().position(|x| x == card) {
            self.hand.remove(pos);
        }
    }

    pub fn get_score(&self) -> usize {
        self.score
    }

    pub fn add_points(&mut self, points: usize) {
        self.score += points;
    }

    pub fn num_tricks_won(&self) -> usize {
        self.tricks
    }
    
    pub fn win_trick(&mut self) {
        self.tricks += 1;
    }
    
    pub fn highest_card(&self, suit: Suit) -> Option<&Card> {
        self.hand.iter().rev().find(|card| card.suit() == suit)
    }
    
    pub fn get_hand(&self) -> Vec<Card> {
        self.hand.clone()
    }
    
    pub fn reset_tricks(&mut self) {
        self.tricks = 0;
    }
}